presto-main/src/main/java/com/facebook/presto/operator/scalar/MapConstructor.java

```diff
@@ -121,55 +120,52 @@ public BuiltInScalarFunctionImplementation specialize(BoundVariables boundVariab
         MethodHandle keyBlockEquals = compose(keyNativeEquals, nativeValueGetter(keyType), nativeValueGetter(keyType));
         MethodHandle keyIndeterminate = functionAndTypeManager.getBuiltInScalarFunctionImplementation(
                 functionAndTypeManager.resolveOperator(INDETERMINATE, fromTypeSignatures((keyType.getTypeSignature())))).getMethodHandle();
-        MethodHandle instanceFactory = constructorMethodHandle(State.class, MapType.class).bindTo(mapType);
 
         return new BuiltInScalarFunctionImplementation(
                 false,
                 ImmutableList.of(
                         valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
                         valueTypeArgumentProperty(RETURN_NULL_ON_NULL)),
-                METHOD_HANDLE.bindTo(mapType).bindTo(keyBlockEquals).bindTo(keyBlockHashCode).bindTo(keyIndeterminate),
-                Optional.of(instanceFactory));
+                METHOD_HANDLE.bindTo(mapType).bindTo(keyType).bindTo(valueType).bindTo(keyBlockEquals).bindTo(keyBlockHashCode).bindTo(keyIndeterminate),
+                Optional.empty());
     }
 
     @UsedByGeneratedCode
     public static Block createMap(
             MapType mapType,
+            Type keyType,
+            Type valueType,
             MethodHandle keyBlockEqual,
             MethodHandle keyBlockHashCode,
             MethodHandle keyIndeterminate,
-            State state,
             SqlFunctionProperties properties,
             Block keyBlock,
             Block valueBlock)
     {
         checkCondition(keyBlock.getPositionCount() == valueBlock.getPositionCount(), INVALID_FUNCTION_ARGUMENT, "Key and value arrays must be the same length");
-        PageBuilder pageBuilder = state.getPageBuilder();
-        if (pageBuilder.isFull()) {
-            pageBuilder.reset();
-        }
-
-        MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) pageBuilder.getBlockBuilder(0);
+        MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) mapType.createBlockBuilder(null, keyBlock.getPositionCount() * 2);
         BlockBuilder blockBuilder = mapBlockBuilder.beginBlockEntry();
+
         for (int i = 0; i < keyBlock.getPositionCount(); i++) {
             if (keyBlock.isNull(i)) {
-                // close block builder before throwing as we may be in a TRY() call
```

Comment:
Why do we remove this?

Reply:
Because there is no state anymore so the comment doesn't apply.